File: Daily Resources/Day09/DSA - Graphs Solutions/high_score.py
import sys
from collections import deque
input = sys.stdin.readline

INF = -10**15  # large negative number, similar to your C++ INF

n, m = map(int, input().split())
a = [0]*m
b = [0]*m
c = [0]*m

for i in range(m):
    a[i], b[i], c[i] = map(int, input().split())
    a[i] -= 1
    b[i] -= 1

dis = [INF]*n
dis[0] = 0

updatedNodes = []

for i in range(n+1):
    updated = False
    for j in range(m):
        if dis[a[j]] == INF:
            continue
        if dis[b[j]] < dis[a[j]] + c[j]:
            dis[b[j]] = dis[a[j]] + c[j]
            updated = True
            if i == n:
                updatedNodes.append(b[j])
    if not updated:
        break

if not updatedNodes:
    print(dis[n-1])
    sys.exit()

graph = [[] for _ in range(n)]
revGraph = [[] for _ in range(n)]

for i in range(m):
    graph[a[i]].append(b[i])
    revGraph[b[i]].append(a[i])

affected = [False]*n
q = deque()
for node in updatedNodes:
    if not affected[node]:
        affected[node] = True
        q.append(node)

while q:
    u = q.popleft()
    for v in graph[u]:
        if not affected[v]:
            affected[v] = True
            q.append(v)

reachableFromN = [False]*n
q2 = deque([n-1])
reachableFromN[n-1] = True

cycleAffectsN = False
while q2:
    u = q2.popleft()
    if affected[u]:
        cycleAffectsN = True
        break
    for v in revGraph[u]:
        if not reachableFromN[v]:
            reachableFromN[v] = True
            q2.append(v)

if cycleAffectsN:
    print(-1)
else:
    print(dis[n-1])

File: Daily Resources/Day09/DSA - Graphs Solutions/high_score.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long
 
int n, m;
vector<int> dis;
int INF = -100000000000000;
vector<int> a, b, c;
 
signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
 
    cin >> n >> m;
    dis.assign(n, INF);
    a.resize(m);
    b.resize(m);
    c.resize(m);
 
    for (int i = 0; i < m; i++)
    {
        cin >> a[i] >> b[i] >> c[i];
        a[i]--; b[i]--;
    }
 
    dis[0] = 0;
 
    vector<int> updatedNodes; // nodes updated on nth iteration
 
    for (int i = 0; i <= n; i++)
    {
        bool updated = false;
        for (int j = 0; j < m; j++)
        {
            if (dis[a[j]] == INF) continue;
            if (dis[b[j]] < dis[a[j]] + c[j])
            {
                dis[b[j]] = dis[a[j]] + c[j];
                updated = true;
                if (i == n)
                    updatedNodes.push_back(b[j]);
            }
        }
        if (!updated) break;
    }
 
    if (updatedNodes.empty()) {
        // No positive cycle affecting any node, print answer directly
        cout << dis[n - 1] << "\n";
        return 0;
    }
 
    // Build graph and reverse graph
    vector<vector<int>> graph(n);
    vector<vector<int>> revGraph(n);
    for (int i = 0; i < m; i++)
    {
        graph[a[i]].push_back(b[i]);
        revGraph[b[i]].push_back(a[i]);
    }
 
    // BFS from updatedNodes (nodes affected by positive cycles)
    vector<bool> affected(n, false);
    queue<int> q;
    for (int node : updatedNodes)
    {
        if (!affected[node])
        {
            affected[node] = true;
            q.push(node);
        }
    }
    while (!q.empty())
    {
        int u = q.front(); q.pop();
        for (int v : graph[u])
        {
            if (!affected[v])
            {
                affected[v] = true;
                q.push(v);
            }
        }
    }
 
    // BFS backwards from n-1 to see if we can reach any affected node
    vector<bool> reachableFromN(n, false);
    queue<int> q2;
    q2.push(n - 1);
    reachableFromN[n - 1] = true;
 
    bool cycleAffectsN = false;
    while (!q2.empty())
    {
        int u = q2.front(); q2.pop();
        if (affected[u])
        {
            cycleAffectsN = true;
            break;
        }
        for (int v : revGraph[u])
        {
            if (!reachableFromN[v])
            {
                reachableFromN[v] = true;
                q2.push(v);
            }
        }
    }
 
    if (cycleAffectsN)
        cout << -1 << "\n";
    else
        cout << dis[n - 1] << "\n";
 
    return 0;
}